### PR TITLE
ENG-7625 Added /deployment/users GET, POST, PUT, and DELETE

### PR DESCRIPTION
**Pull request by dweiss88**

### CODE REVIEW

File: src/frontend/org/voltdb/utils/HTTPAdminListener.java [hunk 1]
```diff
@@ -494,6 +544,212 @@ public void handleUpdateDeployment(String jsonp, String target,
...
+                           throws IOException, ServletException {
+            try {
+                DeploymentType newDeployment = CatalogUtil.getDeployment(new ByteArrayInputStream(getDeploymentBytes()));
+                User user = null;
+                String[] splitTarget = target.split("/");
+                if (splitTarget.length == 3) {
+                    user = findUser(splitTarget[2], newDeployment);
+                }
+                if (user == null) {
+                    response.setStatus(HttpServletResponse.SC_GONE);
```
Comment (dweiss88) [inline src/frontend/org/voltdb/utils/HTTPAdminListener.java]: Change GONE to NOT_FOUND

File: src/frontend/org/voltdb/utils/HTTPAdminListener.java [hunk 2]
```diff
@@ -494,6 +544,212 @@ public void handleUpdateDeployment(String jsonp, String target,
...
         }
+
+        //Handle POST for users
+        public void handleUpdateUser(String jsonp, String target,
+                           Request baseRequest,
+                           HttpServletRequest request,
+                           HttpServletResponse response, Client client)
+                           throws IOException, ServletException {
+            String update = request.getParameter("user");
+            if (update == null || update.length() == 0) {
```
Comment (vtkstef) [inline src/frontend/org/voltdb/utils/HTTPAdminListener.java]: do we support user names that are all spaces? if not then use update.trim().length() instead

File: src/frontend/org/voltdb/utils/HTTPAdminListener.java [hunk 3]
```diff
@@ -494,6 +544,212 @@ public void handleUpdateDeployment(String jsonp, String target,
...
+                user.setPlaintext(newUser.isPlaintext());
+                user.setRoles(newUser.getRoles());
+
+                String dep = CatalogUtil.getDeployment(newDeployment);
+                if (dep == null || dep.trim().length() <= 0) {
+                    response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+                    response.getWriter().print(buildClientResponse(jsonp, ClientResponse.UNEXPECTED_FAILURE, "Failed to build deployment information."));
+                    return;
+                }
+                Object[] params = new Object[2];
```
Comment (vtkstef) [inline src/frontend/org/voltdb/utils/HTTPAdminListener.java]:
```java
Object[] params = new Object[] { null, dep};
```

File: src/frontend/org/voltdb/utils/HTTPAdminListener.java [hunk 4]
```diff
@@ -494,6 +544,212 @@ public void handleUpdateDeployment(String jsonp, String target,
...
+                params[0] = null;
+                params[1] = dep;
+                //Call sync as nothing else can happen when this is going on.
+                client.callProcedure("@UpdateApplicationCatalog", params);
+                response.getWriter().print(buildClientResponse(jsonp, ClientResponse.SUCCESS, "User Updated."));
+                notifyOfCatalogUpdate();
+            } catch (Exception ex) {
+                logger.error("Failed to update user from API", ex);
+                response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+                response.getWriter().print(buildClientResponse(jsonp, ClientResponse.UNEXPECTED_FAILURE, ex.toString()));
```
Comment (vtkstef) [inline src/frontend/org/voltdb/utils/HTTPAdminListener.java]: wouldn't it be better to use guava throwables  to get the full stack trace as a string ? or see if the exception as a printStackTrace() that accepts a Writer as a parameter?

File: src/frontend/org/voltdb/utils/HTTPAdminListener.java [hunk 5]
```diff
@@ -494,6 +544,212 @@ public void handleUpdateDeployment(String jsonp, String target,
...
+                    response.setStatus(HttpServletResponse.SC_OK);
+                }
+
+                String dep = CatalogUtil.getDeployment(newDeployment);
+                if (dep == null || dep.trim().length() <= 0) {
+                    response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+                    response.getWriter().print(buildClientResponse(jsonp, ClientResponse.UNEXPECTED_FAILURE, "Failed to build deployment information."));
+                    return;
+                }
+                Object[] params = new Object[2];
```
Comment (vtkstef) [inline src/frontend/org/voltdb/utils/HTTPAdminListener.java]: see above

File: src/frontend/org/voltdb/utils/HTTPAdminListener.java [hunk 6]
```diff
@@ -494,6 +544,212 @@ public void handleUpdateDeployment(String jsonp, String target,
...
+                params[0] = null;
+                params[1] = dep;
+                //Call sync as nothing else can happen when this is going on.
+                client.callProcedure("@UpdateApplicationCatalog", params);
+                response.getWriter().print(buildClientResponse(jsonp, ClientResponse.SUCCESS, returnString));
+                notifyOfCatalogUpdate();
+            } catch (Exception ex) {
+                logger.error("Failed to create user from API", ex);
+                response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+                response.getWriter().print(buildClientResponse(jsonp, ClientResponse.UNEXPECTED_FAILURE, ex.toString()));
```
Comment (vtkstef) [inline src/frontend/org/voltdb/utils/HTTPAdminListener.java]: see above

File: src/frontend/org/voltdb/utils/HTTPAdminListener.java [hunk 7]
```diff
@@ -494,6 +544,212 @@ public void handleUpdateDeployment(String jsonp, String target,
...
+                if (splitTarget.length == 3) {
+                    user = findUser(splitTarget[2], newDeployment);
+                }
+                if (user == null) {
+                    response.setStatus(HttpServletResponse.SC_GONE);
+                    response.getWriter().print(buildClientResponse(jsonp, ClientResponse.UNEXPECTED_FAILURE, "User not found"));
+                    return;
+                }
+                if (newDeployment.getUsers().getUser().size() == 1) {
+                    newDeployment.setUsers(null);
```
Comment (vtkstef) [inline src/frontend/org/voltdb/utils/HTTPAdminListener.java]: what about all that logic that checks for unwarranted user removals? sole admin ...

File: tests/frontend/org/voltdb/TestJSONInterface.java [hunk 8]
```diff
@@ -1569,6 +1579,84 @@ public void testDeploymentSecurityAuthorizationBasic() throws Exception {
...
+
+            File schemaFile = VoltProjectBuilder.writeStringToTempFile(simpleSchema);
+            String schemaPath = schemaFile.getPath();
+            schemaPath = URLEncoder.encode(schemaPath, "UTF-8");
+
+            VoltProjectBuilder builder = new VoltProjectBuilder();
+            builder.addSchema(schemaPath);
+            builder.addPartitionInfo("foo", "bar");
+            builder.addProcedures(DelayProc.class);
+            builder.setHTTPDPort(8095);
```
Comment (vtkstef) [inline tests/frontend/org/voltdb/TestJSONInterface.java]: maybe use a more wakier port number so it'd be less likely to incur conflicts

**Pull request merged**